pkg/arvo/app/chat-cli.hoon

```diff
@@ -779,37 +641,29 @@
       %+  stag  (crip q.tub)
       wide:(vang & [&1:% &2:% (scot %da now.bowl) |3:%])
     --
-  ::  +obey: apply result
-  ::
-  ::    called upon hitting return in the prompt.
-  ::    if input is invalid, +slug is called.
-  ::    otherwise, the appropriate work is done and
-  ::    the command (if any) gets echoed to the user.
+  ::  +tab-list: command descriptions
   ::
-  ++  obey
-    ^-  outward
-    =+  buf=buf.cli-state
-    =+  fix=(sanity [%nop ~] buf)
-    ?^  lit.fix
-      (slug fix)
-    =+  jub=(rust (tufa buf) read)
-    ?~  jub
-      :_  state
-      [cli-state [(effect:sh-out %bel ~)]~]
-    =^  cal  cli-state  (~(transmit sole-lib cli-state) [%set ~])
-    =^  cards  state  (work u.jub)
-    :_  state
-    :-  cli-state
-    %+  weld
-      ^-  (list card)
-      ::  echo commands into scrollback
-      ?.  =(`0 (find ";" buf))  ~
-      [(note:sh-out (tufa `(list @)`buf)) ~]
-    :_  cards
-    %+  effect-to:sh-out  sole-id
-    :-  %mor
-    :~  [%nex ~]
-        [%det cal]
+  ++  tab-list
```

Comment:
I wonder if we could annotate the parser with tab completion instructions and then get nicer tab completion, e.g. completing `;leave ~dopzod/urbi`

Reply:
> annotate the parser

Wonder if you have a concrete idea of what this'd look like.

Applications could also dynamically generate the `+tab-list`, but that could result in very large suggestion sets. Maybe this should be a nested list instead?

(I don't care too deeply about improving this within this PR, but happy to facilitate future experimentation/development. (: )

Reply:
I'd expose `on-tab` as a `$~([sole-id=@ta buf=(list @c) pos=@u] to-send=tape)` hook more generically, and have tab-list based handling be a library arm (or even a chat-specific arm, since dojo probably _would_ want to dynamically scry on `:app|` or w/e)

~~also `ost=bone` should be in the bowl instead of having to thread ids everywhere~~ I realize this ship has sailed

Reply:
As for the mechanics of completion-oriented parsers - I suspect at some point we'll want a set of extended parser combinators, which deal in `(pair (unit thing-you-wanted) (list highlightable-token))` as opposed to just `(unit thing-you-wanted)`(or rather, as a further refinement/variant on `edge`). For LSP if anything, esp if we can run a pure-function highlighter without urbit's persistence latency constraints 😉.

Or even some intermediate structures, "well the cursor is on the second of three arguments for a generator that wants a `@da` there"; this may be just a dream ofc. It would be nice to have tab close all open parens whatever shape they may be, at least.